src/typecheck/mod.rs

```diff
@@ -842,7 +842,8 @@ fn walk<L: Linearizer>(
                 ctxt.type_env.insert(*id, mk_uniftype::dynamic());
             }
 
-            let pattern_ty = destructuring::build_pattern_type_walk_mode(state, pat);
+            let pattern_ty = destructuring::build_pattern_type_walk_mode(state, &ctxt, pat)
+                .map_err(|e| e.into_typecheck_err(state, rt.pos))?;
```

Comment:
I'll double check the exact positioning here, but I _think_ the subterm positions were "correct" when I tested this out, in that they always point to the record being destructured (which imo is the right thing to do, because the issue is that the entire record value is not compatible with the pattern).

Reply:
I thought that if a unification error happened during `build_pattern` (and not a later checking or unifying), it's rather than the _pattern_ is wrong, as being inconsistent (having a conflicting type annotation for a parent record and a subfield)? In that case the record being destructured can't do much to fix the type error, I think.

Reply:
Ah, my bad, you're totally right. Mutually incompatible type annotations in patterns currently result in the wrong term being highlighted. I'll fix that before I merge this.

Reply:
Note that now we have positions inside `Types`. Might be helpful. Don't hesitate to add a new error variant (or make `build_pattern` return a wrapper around a unification error) if needed to convey the position of the conflicting type (the outer one), as this information should be easily accessible.

Reply:
I ended up just calling `into_typecheck_err` inside `build_pattern_type` using the `pos` from `Types`, which seems to work quite nicely. I also included some snapshot tests for the errors.